internal/pipeline/executor.go

```diff
@@ -288,33 +245,35 @@ func hydrate(p *pkg, hctx *hydrationContext) (resources []*yaml.RNode, err error
 			&kio.PackageBuffer{Nodes: input},
 		},
 		Filters: filters,
-		Outputs: []kio.Writer{pkgWriter, output}, // here may be we don't want to write to the fs yet
+		Outputs: []kio.Writer{output},
 	}
 	err = kioPipeline.Execute()
 	if err != nil {
-		err = fmt.Errorf("failed to run pipeline for pkg %s %w", p.path, err)
-		return
+		err = fmt.Errorf("failed to run pipeline for pkg %v %w", pn.Path(), err)
+		return nil, err
 	}
+	return output.Nodes, nil
 
-	// pkg is hydrated
-	resources = output.Nodes
-	// mark the pkg as wet and update the resources
-	hctx.hydrationStates[p.Path()] = Wet
-	p.resources = resources
-	return resources, err
 }
 
-// filterMetaData filters kpt metadata files such as pipeline,
-// Kptfile, permissions etc.
-func filterMetaData(resources []*yaml.RNode) []*yaml.RNode {
-	var filtered []*yaml.RNode
+// adjustRelPath updates the resources with given relative path.
+func adjustRelPath(resources []*yaml.RNode, relPath string) ([]*yaml.RNode, error) {
 	for _, r := range resources {
-		meta, _ := r.GetMeta()
-		if !strings.Contains(meta.APIVersion, "kpt.dev") {
-			filtered = append(filtered, r)
+		meta, err := r.GetMeta()
+		if err != nil {
+			return resources, err
+		}
+		// TODO(droot): revisit this. There is an edge case where if relativePath(root, curr)
```

Comment:
Worked through an example today and verified that this edge case doesn't exist. resources will be stamped with correct relative path (root --> pkg) when they are either read or generated at that package. Subsequent call to `adjustRelPath` up in the chain (as we move from inward pkg to outer) will always have relative path prefixed. This use-case is exercised using two e2e test cases that we have `pkg-with-generator` and `pkg-with-subpkgs`, so we are good from testing standpoint.

I was initially thinking of using `pkgbuilder` under testutil to implement some more unit tests but that pkg requires more tweaking because it uses older `kptfile` format, so abandoned that idea.